src/main/scala/com/databricks/spark/redshift/RedshiftRelation.scala

```diff
@@ -0,0 +1,131 @@
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.jdbc.JDBCWrapper
+import org.apache.spark.sql.sources._
+import org.apache.spark.sql.types._
+import org.apache.spark.sql.{DataFrame, Row, SQLContext}
+
+/**
+ * Data Source API implementation for Amazon Redshift database tables
+ */
+private [redshift]
+case class RedshiftRelation(jdbcWrapper: JDBCWrapper,
+                            params: MergedParameters,
+                            userSchema: Option[StructType])
+                           (@transient val sqlContext: SQLContext)
+  extends BaseRelation
+  with TableScan
+  with PrunedScan
+  with PrunedFilteredScan
+  with InsertableRelation
+  with Logging {
+
+  override def schema = {
+    userSchema match {
+      case Some(schema) => schema
+      case None => {
+        jdbcWrapper.registerDriver(params.jdbcDriver)
+        jdbcWrapper.resolveTable(params.jdbcUrl, params.table, new Properties())
+      }
+    }
+  }
+
+  override def buildScan(): RDD[Row] = {
+    unloadToTemp()
+    makeRdd(schema)
+  }
+
+  override def buildScan(requiredColumns: Array[String]): RDD[Row] = {
+    val columns = columnList(requiredColumns)
+    unloadToTemp(columns)
+    makeRdd(pruneSchema(schema, requiredColumns))
+  }
+
+  override def buildScan(requiredColumns: Array[String], filters: Array[Filter]): RDD[Row] = {
+    val columns = columnList(requiredColumns)
+    val whereClause = buildWhereClause(filters)
+    unloadToTemp(columns, whereClause)
+    makeRdd(pruneSchema(schema, requiredColumns))
+  }
+
+  override def insert(data: DataFrame, overwrite: Boolean): Unit = {
+    val updatedParams = Parameters.mergeParameters(params.parameters updated ("overwrite", overwrite.toString))
+    new RedshiftWriter(jdbcWrapper).saveToRedshift(sqlContext, data, updatedParams)
+  }
+
+  def unloadToTemp(columnList: String = "*", whereClause: String = ""): Unit = {
+    val conn = jdbcWrapper.getConnector(params.jdbcDriver, params.jdbcUrl, new Properties()).apply()
+    val unloadSql = unloadStmnt(columnList, whereClause)
+    val statement = conn.prepareStatement(unloadSql)
+
+    statement.execute()
+    conn.close()
+  }
+
+  def unloadStmnt(columnList: String, whereClause: String) : String = {
+    val credsString = params.credentialsString()
+    val query = s"SELECT $columnList FROM ${params.table} $whereClause"
+    val fixedUrl = Utils.fixS3Url(params.tempPath)
+
+    s"UNLOAD ('$query') TO '$fixedUrl' WITH CREDENTIALS '$credsString' ESCAPE ALLOWOVERWRITE"
+  }
+
+  def makeRdd(schema: StructType): RDD[Row] = {
+    val sc = sqlContext.sparkContext
+    val rdd = sc.newAPIHadoopFile(params.tempPath, classOf[RedshiftInputFormat],
+      classOf[java.lang.Long], classOf[Array[String]], sc.hadoopConfiguration)
+    rdd.values.map(Conversions.rowConverter(schema))
+  }
+
+  def pruneSchema(schema: StructType, columns: Array[String]): StructType = {
+    val fieldMap = Map(schema.fields map { x => x.metadata.getString("name") -> x }: _*)
```

Comment:
Good question, I don't really get that either. I actually lifted this from the JDBC code because it was all private and I needed the exact same functionality. I've changed to using `x.name` directly and it's fine.